packages/core/echo/live-object/src/ref/ref-impl.ts

```diff
@@ -0,0 +1,140 @@
+export const makeRef = <T extends BaseObject>(obj: T): Ref<T> => {
+  // TODO(dmaretskyi): Extract to `getObjectDXN` function.
+  const id = obj.id;
+  invariant(ObjectId.isValid(id), 'Invalid object ID');
+  const dxn = Reference.localObjectReference(id).toDXN();
+
+  return new RefImpl(dxn, obj);
+};
+
+/**
+ * Constructs a reference that points to the object specified by the provided DXN.
+ */
+// TODO(dmaretskyi): Should be `Ref.fromDXN` but that's not possible because of the circular dependency.
+export const refFromDXN = (dxn: DXN): Ref<any> => {
+  return new RefImpl(dxn);
+};
+
+export interface RefResolver {
+  /**
+   * Resolve ref synchronously from the objects in the working set.
+   *
+   * @param dxn
+   * @param load If true the resolver should attempt to load the object from disk.
+   * @param onLoad Callback to call when the object is loaded.
+   */
+  resolveSync(dxn: DXN, load: boolean, onLoad?: () => void): BaseObject | undefined;
+
+  /**
+   * Resolver ref asynchronously.
+   */
+  resolve(dxn: DXN): Promise<BaseObject | undefined>;
+}
+
+export class RefImpl<T> implements Ref<T> {
+  #dxn: DXN;
+  #resolver?: RefResolver = undefined;
+  #signal = compositeRuntime.createSignal();
+
+  /**
+   * Target is set when the reference is created from a specific object.
+   * In this case, the target might not be in the database.
+   */
+  #target: T | undefined = undefined;
+
+  /**
+   * Callback to issue a reactive notification when object is resolved.
+   */
+  #resolverCallback = () => {
+    this.#signal.notifyWrite();
+  };
+
+  constructor(dxn: DXN, target?: T) {
+    this.#dxn = dxn;
+    this.#target = target;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  get dxn(): DXN {
+    return this.#dxn;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  get target(): T | undefined {
+    invariant(this.#resolver, 'Resolver is not set');
+    this.#signal.notifyRead();
+
+    return this.#resolver.resolveSync(this.#dxn, true, this.#resolverCallback) as T | undefined;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  async load(): Promise<T> {
+    invariant(this.#resolver, 'Resolver is not set');
+    const obj = await this.#resolver.resolve(this.#dxn);
+    if (obj == null) {
+      throw new Error('Object not found');
+    }
+    return obj as T;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  async tryLoad(): Promise<T | undefined> {
+    invariant(this.#resolver, 'Resolver is not set');
+    return (await this.#resolver.resolve(this.#dxn)) as T | undefined;
+  }
+
+  /**
+   * Internal method to set the resolver.
+   * @internal
+   */
+  _setResolver(resolver: RefResolver) {
+    this.#resolver = resolver;
+  }
+
+  /**
+   * Internal method to get the saved target.
+   * Not the same as `target` which is resolved from the resolver.
+   * @internal
+   */
+  _getSavedTarget() {
+    return this.#target;
+  }
+}
+
+/**
+ * Internal API for setting the reference resolver.
```

Comment:
`@internal`

Reply:
can't because internal will make it package-private

Reply:
perhaps worth changing the function name to make it more apparent, e.g.: `__setRefResolver`